README.md

```diff
@@ -11,4 +11,73 @@ However, if exceptions occur frequently, consider adjusting guidelines (by creat
 
 ## Table of contents
 
-// TODO
+- [Git](#git)
+  - [Workflow](#workflow)
+  - [Commit messages](#commit-messages)
+
+## [Git](#git)
+
+### [Workflow](#workflow)
+
+1. Create a feature branch  
+
+    - Prefix branch name with `feature/` or `fix/` when applicable; otherwise, prefix with your initials, e.g. `tk/`
+    - Use kebabcase
+    - Include the issue number when applicable
+    - Briefly describe the change  
+
+    Examples: `feature/610-seelog-wrapper`, `fix/tequilapi-startup`, `tk/polish`
+
+2. Implement the changes (series of commits)
+    - It's best when commmits are clearly separated, i.e. each fix, change, refactoring, extra tests go into their own commits
+    - When fixing up a commit that already exists on your feature branch, try your best to squash the fix to the related commit so that we can avoid noise in git history
+      - If it's the last commit:
+
+        ```
+        $ git commit --amend
+        ```
+
+      - Otherwise, do [an interactive rebase](https://help.github.com/en/articles/about-git-rebase) to squash them into cohesive commits with good messages:
+
+        ```
+        $ git fetch origin
+        $ git rebase -i origin/master
+        ```
+3. Rebase frequently to incorporate upstream changes
+
+    ```
+    $ git pull --rebase origin master
+    ```
+
+4. Push your branch and create a pull request (PR)
+    - Request reviews as suggested and anyone else you see fit
+    - Wait for approvals and CI checks to pass
+
+5. Merge PR and remove the branch (or ask someone with push rights to do it)
+
+General tips:
+
+- Try to integrate with master branch ASAP. Prefer delivering feature in smaller working chunks rather than one huge PR.
+- Avoid merging `origin/master` into feature branch. Rebase instead to keep git history as linear as possible.
```

Comment:
Since we require all PR branches to be up to date with `master`, you often have to update your branch multiple times until it gets merged. After you already have approvals, rebasing it multiple times and getting approvals again is time-consuming for all and delays merging. So we usually are ok with creating merge commits for this case using web interface to avoid dismissing approvals.

Reply:
We should then agree on SQUASHING to master...

Reply:
There is another option: don't worry about perfectly cleaned git history. :)

Reply:
Some PRs contains many commits and squashing it makes a single huge commit. Later it's starting to be harder to navigate through history with huge commits.

Reply:
not seeking for perfection, but seeing 7 merges in a row is a bit baffling.

Reply:
All approaches has pluses and minuses - personally I dislike squashing to master, because you get huge commits where all changes are merged, so it's hard to understand them and you can't revert single smaller commits.

I prefer rebase'ing, but merge commits are fine once you have a PR - it makes easier to review and saves time by keeping approvals.